frontend/src/shape/Shape.ts

```diff
@@ -0,0 +1,106 @@
+// used to define the shape of data returned from the API
+// not the most elegant solution but so long as we're careful, it should be fine
+// inb4 "should be replaced with graph-ql", you would not be wrong
+
+import { ArrayElement } from "../magic/ArrayElement";
+
+export abstract class Shape<A> {
+    protected abstract throwIfMisfit(obj: any): void;
+
+    public assert(obj: any): A {
+        try {
+            this.throwIfMisfit(obj);
+            return obj as A;
+        } catch (e) {
+            if (!(e instanceof InvalidShapeError)) {
+                throw new InvalidShapeError();
+            } else {
+                throw e;
+            }
+        }
+    }
+}
+
+export type InvalidShapeErrorProblem =
+    | "KEY"
+    | "TYPE"
+    | "EQUALITY"
+    | "INCLUSION";
+export class InvalidShapeError extends Error {
+    constructor(
+        problemType: InvalidShapeErrorProblem | null = null,
+        actual: any = "",
+        expected: any = "",
+    ) {
+        let description;
+
+        switch (problemType) {
+            default:
+                description = "";
+            case "KEY":
+                description = `Could not find expected key: '${expected}'`;
+            case "TYPE":
+                description = `Mismatch type, expected: '${expected}', got: '${actual}'`;
+            case "EQUALITY":
+                description = `Unequal values, expected: '${expected}', got: '${actual}`;
+            case "INCLUSION":
+                description = `Expected value ${actual} to be an element of [${expected}]`;
+        }
+
+        super(`SHAPE_ERROR: ${description}`);
+    }
+}
+
+export function assertKeyInShape(object: any, key: string) {
+    if (!object.hasOwnProperty(key)) {
+        throw new InvalidShapeError("KEY", undefined, key);
+    }
+}
+
+// note: you can't refactor these two lists into a single list
+// the const "string" is different from the type "string"
+// every javascript type has a string/literal copy in typescript
+export type KnownJavascriptType = [
+    "string",
+    "number",
+    "bigint",
+    "boolean",
+    "symbol",
+    "undefined",
+    "object",
+    "function",
+];
+export const KnownJavascriptTypes: KnownJavascriptType = [
+    "string",
+    "number",
+    "bigint",
+    "boolean",
+    "symbol",
+    "undefined",
+    "object",
+    "function",
+];
+
+export function assertType<A extends ArrayElement<KnownJavascriptType>>(
```

Comment:
I am of the belief that this entire class is essentially what we had before in the big ass GraphQL folder with all the types. But now it was written using your time and effort rather than crafted by some god-given library. Either way what's done is done.

Reply:
Call me stubborn but I'd rather be deeply knowledgeable about a library that I wrote to solve a problem than relying on setting up some other library with its own quirks and documentation I have to read. Plus it's easier to modify and expand to fit our needs. And a god-given library would implement some 90 other functions we wouldn't need, so our library is more streamlined and optimized, tell that to the shareholders. COPE